### PR TITLE
Replace Overriden getTriggerPhrase() with setTriggerPhrase() usage

### PR DESCRIPTION
**Pull request by Alex-Vasile**
Whenever an ability triggers and goes on the stack it will generate a Human-readable rule using `.getRule()`. This is the first of several PRs to speed this up.

This PR removes uses of `getTriggerPhrase()` in favour of the existing `setTriggerPhrase()` functionality.

The next step is to look at memoizing the results at `getRule()` so it doesn't thrash through as much memory.

For #9302.

### CODE REVIEW

**Comment by awjackson** [comment]
I understand the first two commits here, but you lost me at the third. Why does every single `TriggeredAbility` class now need a new static String field *and* an override method that always does exactly the same thing for every class (i.e. return that class's particular static String)? This just seems like adding additional boilerplate to do the same thing as before. What am I missing?

**Comment by Alex-Vasile** [comment]
> I understand the first two commits here, but you lost me at the third. Why does every single `TriggeredAbility` class now need a new static String field *and* an override method that always does exactly the same thing for every class (i.e. return that class's particular static String)? This just seems like adding additional boilerplate to do the same thing as before. What am I missing?

The reason for the boiler plate is that the static Sting *cannot* go into the superclass (TriggeredAbilityImpl) since then all triggered abilities would share that same string. Each subclass needs it's own separate static string.

One alternative would be to have the superclass contain a static dictionary which stores a string the static string for each class.

But I liked the first approach more since it means we don't have to worry about that dictionary getting big, or about concurrently modifying it.

**Comment by awjackson** [comment]
What was wrong with your initial approach, where every class just called `setTriggerPhrase` in its constructor?

**Comment by Alex-Vasile** [comment]
> What was wrong with your initial approach, where every class just called `setTriggerPhrase` in its constructor?

That will still generate and store a separate string per instance of the class.

Either one will work, but this one will theoretically have a lower ram usage since there is only one copy of the string.

**Comment by awjackson** [comment]
> > What was wrong with your initial approach, where every class just called `setTriggerPhrase` in its constructor?
> 
> That will still generate and store a separate string per instance of the class.

It should only generate a new string per instance if the argument to `setTriggerPhrase` is an expression (e.g. "Whenever " + filter.getMessage() + " attacks"). If the argument to `setTriggerPhrase` is simply a string literal, then every instance of the class should get a reference to the same string. Strings in Java are immutable (string expressions like the one above are in fact syntactic sugar for `StringBuilder`) and are automatically "interned" by the compiler/runtime.

**Comment by Alex-Vasile** [comment]
> > > What was wrong with your initial approach, where every class just called `setTriggerPhrase` in its constructor?
> > 
> > 
> > That will still generate and store a separate string per instance of the class.
> 
> It should only generate a new string per instance if the argument to `setTriggerPhrase` is an expression (e.g. "Whenever " + filter.getMessage() + " attacks"). If the argument to `setTriggerPhrase` is simply a string literal, then every instance of the class should get a reference to the same string. Strings in Java are immutable (string expressions like the one above are in fact syntactic sugar for `StringBuilder`) and are automatically "interned" by the compiler/runtime.

You're saying the JVM will cache the string literal and point each new instance of the class to the same memory space?
Alright. I'll test it out to confirm it to myself, and if it there isn't a memory difference I'll get rid of the static strings so that each class uses the same approach.

**Comment by awjackson** [comment]
> You're saying the JVM will cache the string literal and point each new instance of the class to the same memory space?

Yes, this behaviour is in fact part of the Java language specification: https://docs.oracle.com/javase/specs/jls/se7/html/jls-15.html#jls-15.28

>Compile-time constant expressions of type String are always "interned" so as to share unique instances

**Comment by Alex-Vasile** [comment]
> > You're saying the JVM will cache the string literal and point each new instance of the class to the same memory space?
> 
> Yes, this behaviour is in fact part of the Java language specification: https://docs.oracle.com/javase/specs/jls/se8/html/jls-15.html#jls-15.28
> 
> > Compile-time constant expressions of type String are always "interned" so as to share unique instances

I'll get rid of the static strings

**Comment by Alex-Vasile** [comment]
Alright. Finished all but one of the abilities. I'm going to leave that one since that will require much more refactoring.

The current approach of telescoping constructors is really getting out of hand with that class. But addressing that is beyond the scope of this PR.

File: Mage/src/main/java/mage/abilities/common/DealsDamageToOpponentTriggeredAbility.java [hunk 1]
```diff
@@ -32,6 +32,7 @@ public DealsDamageToOpponentTriggeredAbility(Effect effect, boolean optional, bo
         super(Zone.BATTLEFIELD, effect, optional);
         this.onlyCombat = onlyCombat;
         this.setTargetPointer = setTargetPointer;
+        setTriggerPhrase("Whenever {this} deals " + (onlyCombat ? "combat" : "") + " damage to an opponent, ");
```
Comment (awjackson) [inline Mage/src/main/java/mage/abilities/common/DealsDamageToOpponentTriggeredAbility.java:35]: This generates an extra space when `onlyCombat` is false. The space that's before "damage" needs to be moved to after "combat", matching other similar trigger phrases.

File: Mage/src/main/java/mage/abilities/common/SpellCastControllerTriggeredAbility.java [hunk 2]
```diff
@@ -15,7 +15,7 @@
  */
 public class SpellCastControllerTriggeredAbility extends TriggeredAbilityImpl {
 
-    protected FilterSpell filter;
+    protected FilterSpell filter; // TODO: This sould be final, but is not because of the telescoping contructors
```
Comment (awjackson) [inline Mage/src/main/java/mage/abilities/common/SpellCastControllerTriggeredAbility.java:18]: What is blocking this from being made final? Is it just `ProwessAbility`?

File: Mage/src/main/java/mage/abilities/keyword/BountyAbility.java [hunk 3]
```diff
@@ -20,15 +20,16 @@ public class BountyAbility extends DiesCreatureTriggeredAbility {
...
     }
 
     public BountyAbility(Effect effect, boolean optional) {
-        super(effect, optional, bountyCounterFilter);
+        this(effect, optional, false);
     }
 
     public BountyAbility(Effect effect, boolean optional, boolean setTargetPointer) {
         super(effect, optional, bountyCounterFilter, setTargetPointer);
+        setTriggerPhrase("<i>Bounty</i> &mdash; ");
```
Comment (awjackson) [inline Mage/src/main/java/mage/abilities/keyword/BountyAbility.java:32]: Looking at the cards that use this ability (they're all from the fanmade Star Wars set) it seems the trigger phrase is actually supposed to be "<i>Bounty</i> &mdash; Whenever a creature an opponent controls with a bounty counter on it dies, ". All cards with this ability have broken rules text that doesn't actually say what the trigger condition is (I guess whoever made the Star Wars set didn't understand the difference between ability words and keywords...)

**Review by awjackson — APPROVED** [review]

**Comment by Alex-Vasile** [comment]
Going to merge main to fix the conflicts, run the test, and then merge it in

**Pull request merged**